netaddr.go

```diff
@@ -1086,6 +1086,78 @@ func (s *IPSet) RemoveSet(b *IPSet) {
 	}
 }
 
+// extractPrefix will return a Prefix with a bitlength of length which is part of the given Prefix.
+func (p IPPrefix) extractPrefix(length uint8) (IPPrefix, error) {
+	if length <= p.Bits {
+		return IPPrefix{}, fmt.Errorf("length must be greater than prefix.Bits")
+	}
+
+	r := p.Range()
+	subrange := IPRange{From: r.From, To: r.To.Prior()}
+	srpfx := subrange.Prefixes()
+	if len(srpfx) < 2 {
+		return IPPrefix{}, fmt.Errorf("unable to create child prefix for length:%d", length)
+	}
+	for _, srp := range srpfx {
+		if srp.Bits == length {
+			return srp, nil
+		}
+	}
+	return IPPrefix{}, fmt.Errorf("no prefix with length:%d found in %s", length, p)
+}
+
+// ExtractPrefix will return a Prefix with a bitlength of length which is part of the given IPSet.
+// The Prefix is also removed from the IPSet.
+//
+// Unlike RemovePrefix this function will not require you to give a Prefix, instead you
+// can define what size your new Prefix should have, and a more or less random prefix from
+// within the whole address space of this IPSet is taken out and removed.
+//
+// This might be useful if you want to divide a existing network into smaller parts.
+//
+// Example usage:
+// pfx,_ := netaddr.ParsePrefix("192.168.0.0/16")
+// set.AddPrefix(pfx)
+// childPrefix, _ := set.ExtractPrefix(18)
+//
+// childPrefix will be: "192.168.0.0/18"
+func (s *IPSet) ExtractPrefix(length uint8) (IPPrefix, error) {
+	prefixes := s.Prefixes()
+	if len(prefixes) == 0 {
+		return IPPrefix{}, fmt.Errorf("no more child prefixes contained in prefix pool")
+	}
+	existingPrefixes := make(map[uint8]IPPrefix)
+	for _, prefix := range prefixes {
```

Comment:
I think you can make this more efficient:

```
var bestFit IPPrefix
for _, prefix := range prefixes {
  if prefix.Bits > length {
    continue
  }
  if bestFit.IP.IsZero() || prefix.Bits > bestFit.Bits {
    bestFit = prefix
  }
  if bestFit.Bits == length {
    // exact match, done
    break
  }
}
```

At the end of that iteration, bestFit will be the best prefix to use - either an exact match, or the closest in size - or bestFit.IP.IsZero()==true if no suitable prefix exists. As a bonus, it exits early if it finds an exact match, and you can remove the explicit test for len()==0 above.

Reply:
Unsure, don't understand really, bestFit is not set in you example code.

Reply:
It's set in the loop, in the second if. It starts out as the zero value, then gets overridden by the first "big enough" prefix, and gets updated as more precise prefixes are found.